build/ClientAnnotator.php

```diff
@@ -0,0 +1,207 @@
+                        $versionsWithMethod[$supportedVersion] = true;
+                    }
+
+                    if (count(array_filter($versionsWithMethod, 'boolval'))
+                        !== count($versions)
+                    ) {
+                        $supportedIn = implode(
+                            ', ',
+                            array_filter($versionsWithMethod, 'boolval')
+                        );
+                        $docString .= " (present in versions $supportedIn)";
+                    }
+                }
+
+                $annotations []= [
+                    'doc' => $docString,
+                    'return' => '\\Aws\\Result',
+                    'signature' => lcfirst($method) . '(array $args = [])',
+                ];
+            }
+        }
+
+        return $annotations;
+    }
+
+    private function getMethods()
+    {
+        static $methods;
+
+        if (empty($methods)) {
+            $methods = [];
+            $provider = ApiProvider::defaultProvider();
+
+            foreach ($this->getVersions() as $version) {
+                foreach ($provider(
+                    'docs',
+                    $this->getEndpoint(),
+                    $version
+                )['operations'] as $method => $docString) {
+                    if (empty($methods[$method])) {
+                        $methods[$method] = [
+                            'versions' => [],
+                            'docStrings' => [],
+                        ];
+                    }
+
+                    $methods[$method]['versions'] []= $version;
+                    $methods[$method]['docStrings'] []= strip_tags($docString);
+                }
+            }
+        }
+
+        return $methods;
+    }
+
+    private function getVersions()
+    {
+        static $versions;
+
+        if (empty($versions)) {
+            $versions = ApiProvider::defaultProvider()
+                ->getVersions($this->getEndpoint());
+
+            // ensure that versions are always iterated from oldest to newest
+            sort($versions);
+        }
+
+        return $versions;
+    }
+
+    private function getApiDefinition()
+    {
+        static $api;
+
+        if (empty($api)) {
+            $provider = ApiProvider::defaultProvider();
+            $api = $provider('api', $this->getEndpoint(), 'latest');
+        }
+
+        return $api;
+    }
+
+    private function getEndpoint()
+    {
+        static $endpoint;
+
+        if (empty($endpoint)) {
+            $service = strtolower(
+                preg_replace('/Client$/', '', $this->reflection->getShortName())
+            );
+
+            $endpoint = Aws\manifest($service)['endpoint'];
+        }
+
+        return $endpoint;
+    }
+
+    private function getDefaultDocComment()
+    {
+        return <<<EODC
+/**
+ * This client is used to interact with the **{$this->getApiDefinition()['metadata']['serviceFullName']}** service.
```

Comment:
Can this just become something much simpler like: `{$this->getApiDefinition()['metadata']['serviceFullName']} client`